### ISSUE TITLE
Add Support for IBM Notifier Notifications

### ISSUE THREAD

**Issue by froggtech** [Installomator/Installomator#542]
IBM Notifier gives cleaner notification to the end-user than Jamf Helper. Would be nice if Installomator looked for IBM Notifier first, this way admins using an MDM other than Jamf could get better notifications to their users
<img width="573" alt="Screen Shot 2022-05-18 at 7 41 08 AM" src="https://user-images.githubusercontent.com/33664970/169040963-685ff706-0883-4878-b83b-deebf90ba924.png">
.

**Comment by spikehed** [comment]
I have done this for my deployments - but using swiftDialog instead - maybe we can have both options available...

```
displaynotification() { # $1: message $2: title
    message=${1:-"Message"}
    title=${2:-"Notification"}
    manageaction="/usr/local/bin/dialog"

    if [[ -x "$manageaction" ]]; then
         "$manageaction" -m "$message" -t "$title"  -o -p -s -d -i $LOGO --position topright --timer 5
    else
        runAsUser osascript -e "display notification \"$message\" with title \"$title\""
    fi
}
```

**Comment by acodega** [comment]
I'm not sure if we should make a separate issue.. but I think It may be time to rework displayNotification in order to implement additional support. Right now is a good time in the "toolbox" space with apps like swiftDialog and dataJAR Notifier.

I think we should have a notification option defined in the script. If we add support for IBM Notifier, swiftDialog, dataJAR Notifier then one can argue we'll have to make it configurable and not automatically determine what to use, the way it checks for hub or Jamf right now. 

For example, if you have swiftDialog and dataJAR Notifier deployed, you might want to use dataJAR Notifier during some Installomator runs to notify the user for updates and then use swiftDialog when you want to be more forceful.

**Comment by scriptingosx** [comment]
deferring this for later. swiftDialog is implemented in #641. Maybe that can be adapted for IBM Notifier?

**Comment by kryptonit1** [comment]
Support for IBM Notifier would be really nice.

**Comment by Theile** [comment]
Please let us know what exactly you would like to see and how it's better than swiftDialog, and whatever else eill justify the time to implement it, thank you.

**Comment by mks-007** [comment]
It has the added advantage of doing banner notifications along side the messages. An all in one as it were ;-)

**Comment by kryptonit1** [comment]
> It has the added advantage of doing banner notifications along side the messages. An all in one as it were ;-)

I agree with this, but another reason is that if you have already implemented IBM Notifier, then you don’t want to implement another tool that does the same type of things just to use it for Installomator. The same would probably apply to people running swiftDialog if Installomator only had support for IBM Notifier.

**Comment by mactesting12** [comment]
I'll post what I'm using for IBM Notifier later this week. I need to test with 2.9.1 as the alerts/banners didn't work until this version which I need to account for in the Installomator changes.

**Comment by mactesting12** [comment]
Here's what I've edited in Installomator to have IBM Notifier work. One thing to note is when `displaynotification` is called it needs to run as a sub process so I had to search the script for all instances of `displaynotification` and add an `&` to the end of each line. This way Installomator isn't waiting on the status of the notification.

`displaydialog() { # $1: message $2: title
    message=${1:-"Message"}
    title=${2:-"Installomator"}
    notifier="/Applications/IBM Notifier.app/Contents/MacOS/IBM Notifier"
    LOGO="/Library/Application Support/CompanyName/Logos/Logo.png"
    type="popup"
    button_1="Not Now"
    button_2="Quit and Update"
    manageaction="/Library/Application Support/JAMF/bin/jamfHelper.app/Contents/MacOS/jamfHelper"
    # Searches for IBM Notifier first and if found will display the notification in IBM Notifier for 60 seconds otherwise times out to "Quit and Update"
    # If IBM Notifier isn't found, then searches for Jamf Helper and displays the notification in Jamf Helper
    # If neither application is found, use Apple Script
    if [[ -x "$notifier" ]]; then
        CMD_RESULT=$(runAsUser "$notifier" \
            "-type" "$type" \
            "-bar_title" "Installation Notification" \
            "-title" "$title" \
            "-subtitle" "$message" \
            "-logo" "$LOGO" \
            "-main_button_label" "$button_1" \
            "-secondary_button_label" "$button_2" \
            -silent \
            -always_on_top; echo $?)
        if [[ $CMD_RESULT -eq 2 ]]; then
            echo "Quit and Update"
        else
            echo "Not Now"
        fi
    elif [[ -x "$manageaction" ]]; then
        CMD_RESULT=$("$manageaction" -windowType utility -description "$message" -heading "$title" -icon "$LOGO" -button1 "$button_1" -button2 "Quit/Update" -defaultButton 2 -cancelButton 1)
        if [[ $CMD_RESULT -eq 2 ]]; then
            echo "Quit and Update"
        else
            echo "Not Now"
        fi
    else
        runAsUser osascript -e "button returned of (display dialog \"$message\" with  title \"$title\" buttons {\"$button_1\", \"$button_2\"} default button \"$button_2\" with icon POSIX file \"$LOGO\")"
    fi
}

displaydialogContinue() { # $1: message $2: title
    message=${1:-"Message"}
    title=${2:-"Installomator"}
    manageaction="/Applications/IBM Notifier.app/Contents/MacOS/IBM Notifier"
    LOGO="/Library/Application Support/CompanyName/Logos/Logo.png"
    button_1="Quit and Update"
    if [[ -x "$manageaction" ]]; then
        CMD_RESULT=$(runAsUser "$manageaction" \
            "-type" "popup" \
            "-bar_title" "Coalition Notifier" \
            "-title" "$title" \
            "-subtitle" "$message" \
            "-icon_path" "$LOGO" \
            "-main_button_label" "$button_1" \
            -silent \
            -always_on_top; echo $?)
    else
        runAsUser osascript -e "button returned of (display dialog \"$message\" with  title \"$title\" buttons {\"Quit and Update\"} default button \"Quit and Update\" with icon POSIX file \"$LOGO\")"
    fi
}

displaynotification() { # $1: message $2: title
    message=${1:-"Message"}
    title=${2:-"Notification"}
    manageaction="/Applications/IBM Notifier.app/Contents/MacOS/IBM Notifier"
    LOGO="/Library/Application Support/CompanyName/Logos/Logo.png"
    type="banner"

    if [[ -x "$manageaction" ]]; then
        CMD_RESULT=$("$manageaction" \
        "-type" "$type" \
        "-title" "$title" \
        "-subtitle" "$message" \
        -silent)
    else
        runAsUser osascript -e "display notification \"$message\" with title \"$title\""
    fi
}`

**Comment by SMartorelli** [comment]
Hello team, I'm just dropping a message here to understand if the issue is being currently worked on. I just started familiarising with the project because of a request from one of our OS users. Don't you think would be much better to abstract the notification workflow and create a common communication protocol that then enable a team to integrate their own preferred notification tool? same kind of abstraction like you already have with the app labels. Then every team could easily integrate their notification tool and avoid tools redundancy in their environment. 
I would be happy to help with it in case.

**Comment by Theile** [comment]
Fixed in #1046
